### ISSUE TITLE
Mixing of two audio sources

### ISSUE THREAD

**Issue by markusmoenig** [Rodrigodd/audio-engine#3]
Hi,

congrats! You are the first library which works for me both on desktop and web.

The only things I would need is volume settings and mixing of two audio sources (background music and effects). Do you plan to support it ? 

Thanks a lot!

**Comment by Rodrigodd** [comment]
Well, all `SoundSource`s are mixed through [this `Mixer`](https://github.com/Rodrigodd/audio-engine/blob/master/src/lib.rs#L144-L369), but it is private. But you could copy and paste it to your code, and handle the mixing yourself, you would only need to add a `master_volume: f32` field, and multiple its value to each sound volume [here](https://github.com/Rodrigodd/audio-engine/blob/3d0da3711b5cc78e7192d616ebb1d4069920707d/src/lib.rs#L341-L350). And then you would add a mixer for each type of audio source, and you add them to the AudioEngine in a `Arc<Mutex<Mixer>>` (but `SoundSource` is not implemented for `Arc<Mutex<T: SoundSource>>` so you would need to wrap it in newtype) (may be a lot easier to simply fork this repo, and change Mixer directly).

Or you could even use a `group_volume: HashMap<GroupId, f32>` field (group is probably not the right terminology) and each `SoundInner` would hold its `GroupId`, that would be passed as parameter in [add_sound](https://github.com/Rodrigodd/audio-engine/blob/3d0da3711b5cc78e7192d616ebb1d4069920707d/src/lib.rs#L219-L224). This would have better performance and better API, I think.

> Do you plan to support it?

I don't have a plan, I only go implementing things as I see necessity in them. This API that I proposed seems reasonable at first glance, so if I find some free time I may implement that. But PR's are also welcome!

**Comment by markusmoenig** [comment]
Thanks for the info. I am just super busy with my Eldiron project and you know the code much. better than I do. I would be happy to sponsor  this feature a bit ?

**Comment by Rodrigodd** [comment]
I may find time to implement that at the end of the month.

I am not accepting sponsoring.

**Comment by Rodrigodd** [comment]
@markusmoenig I have implemented the sound grouping. The API is displayed below. Can you give it a test? If you don't find any problem, I will go ahead and publish a new version.

I put the sound grouping behind a new API to avoid a major breakage:

```rust
use audio_engine::{AudioEngine, WavDecoder};

#[derive(Eq, Hash, PartialEq)]
enum Group {
    Effect,
    Music,
}

let audio_engine = AudioEngine::with_groups::<Group>()?;
let mut fx = audio_engine.new_sound_with_group(Group::Effect, my_fx)?;
let mut music = audio_engine.new_sound_with_group(Group::Music, my_music)?;

fx.play();
music.play();

// decrease all music volume, for example
audio_engine.set_group_volume(Group::Music, 0.1);
```

**Comment by markusmoenig** [comment]
Oh wow, great, thanks! I will give it a try asap, right now my code is all broken up due to a refactor but will try it asap!!

**Comment by markusmoenig** [comment]
On the desktop it works fine as far as I can test, compiling for the web generates an error:

--> /Users/markusmoenig/.cargo/git/checkouts/audio-engine-cae76c3712fa3f1b/a2f1dc7/src/engine.rs:216:38
214 |     pub fn with_groups<G: Eq + Hash + Send>() -> Result<AudioEngine<G>, &'static str> {
    |                        - this type parameter
215 |         let mixer = Arc::new(Mutex::new(Mixer::<G>::new(2, super::SampleRate(48000))));
216 |         let backend = Backend::start(mixer.clone())?;
    |                                      ^^^^^^^^^^^^^ expected `()`, found type parameter `G`

**Comment by Rodrigodd** [comment]
@markusmoenig I fixed that error with the [last commit](https://github.com/Rodrigodd/audio-engine/commit/5d20e46c8d483626d92bd697ac77556fb0624d49).

**Comment by markusmoenig** [comment]
Working fine now!

**Comment by Rodrigodd** [comment]
Ok, thanks!

So I go ahead and publish a new version.

**Comment by Rodrigodd** [comment]
This feature is implemented in version 0.4.0.
